src/core/annotation.js

```diff
@@ -818,10 +818,15 @@ class WidgetAnnotation extends Annotation {
       "";
     const fieldType = getInheritableProperty({ dict, key: "FT" });
     data.fieldType = isName(fieldType) ? fieldType.name : null;
-    this.fieldResources =
-      getInheritableProperty({ dict, key: "DR" }) ||
-      params.acroForm.get("DR") ||
-      Dict.empty;
+
+    this.fieldResources = Dict.merge({
```

Comment:
In general I'm fine with this change but when saving a form, we copy the resources dictionary:
https://github.com/mozilla/pdf.js/blob/master/src/core/annotation.js#L1040
So at the end it'll increase the file size uselessly, especially if the pdf is well formated.
That lets me think that we don't need to add fieldResources when this one is coming from Acroform...
To avoid that, my idea, when fixing https://github.com/mozilla/pdf.js/issues/12294, was to have a kind of `lazy` merge:
just put in the dict what we really need.

Reply:
> So at the end it'll increase the file size uselessly, especially if the pdf is well formated.

Given that the resources are usually `Ref`s, that should hopefully reduce the size-impact in general; however I do agree that avoiding this if at all possible would be preferable.

> That lets me think that we don't need to add fieldResources when this one is coming from Acroform...

Assuming that you can detect this case easily, and implement it without mutating any *existing* `Dict`s, that could perhaps work (although I'm not entire sure I understand exactly what you have in mind). Please explore this idea further!

(If you do need to merge `Dict`s, with sub-resources, this patch provides a correct way of doing so.)

> To avoid that, my idea, when fixing #12294, was to have a kind of `lazy` merge:
> just put in the dict what we really need.

The problem was with *how* it was implemented, since you unfortunately added special-cases (and dare I say hacks) to the *general* `Dict` abstraction used throughout the code-base to satisfy just *one* very specific use-case.
Hence my objections, since that wasn't really maintainable and also would have meant mutating existing `Dict`s in a way that you absolutely shouldn't do.

---

I'll remove the `src/core/annotation.js` changes from this patch, but I think that it's probably generally useful to have it regardless.

Reply:
> > So at the end it'll increase the file size uselessly, especially if the pdf is well formated.
> 
> Given that the resources are usually `Ref`s, that should hopefully reduce the size-impact in general; however I do agree that avoiding this if at all possible would be preferable.
> 
> > That lets me think that we don't need to add fieldResources when this one is coming from Acroform...
> 
> Assuming that you can detect this case easily, and implement it without mutating any _existing_ `Dict`s, that could perhaps work (although I'm not entire sure I understand exactly what you have in mind). Please explore this idea further!
I'll file a bug for that: my idea is to reduce the file size when saving.

> 
> (If you do need to merge `Dict`s, with sub-resources, this patch provides a _correct_ way of doing so.)
> 
> > To avoid that, my idea, when fixing #12294, was to have a kind of `lazy` merge:
> > just put in the dict what we really need.
> 
> The problem was with _how_ it was implemented, since you unfortunately added special-cases (and dare I say hacks) to the _general_ `Dict` abstraction used throughout the code-base to satisfy just _one_ very specific use-case.
> Hence my objections, since that wasn't really maintainable and also would have meant mutating existing `Dict`s in a way that you absolutely shouldn't do.

Agreed !
Maybe we could create an object LazyMergedDict (extends Dict) which will keep a list of dicts. When we make a `get`, we have a look in each dict until we find it and then we add the key+value in the dict.
Wdyt ? do you have other ideas ?

> 
> I'll remove the `src/core/annotation.js` changes from this patch, but I think that the better `Dict.merge` functionality is probably generally useful to have regardless.